statistics/ma_performance_calculator.py

```diff
@@ -0,0 +1,118 @@
+"""Module responsible for calculating our approach for multi_agent precision and recall."""
+import csv
+from collections import defaultdict
+from pathlib import Path
+from typing import List, Dict, Tuple, Union
+from pddl_plus_parser.lisp_parsers import DomainParser
+from pddl_plus_parser.models import (
+    Domain,
+    Observation,
+    MultiAgentObservation,
+    Operator,
+)
+from statistics import SemanticPerformanceCalculator, _calculate_precision_recall
+from utilities import LearningAlgorithmType
+
+MA_PERFORMANCE_STATS = [
+    "learning_algorithm",
+    "action_name",
+    "num_trajectories",
+    "precondition_precision",
+    "precondition_recall",
+    # "effects_precision",
+    # "effects_recall",
+]
+
+class MaSamPerformanceCalculator(SemanticPerformanceCalculator):
+    """Class responsible for calculating the multi_agent precision and recall of a model."""
+
+    def __init__(
+        self,
+        model_domain: Domain,
+        model_domain_path: Path,
+        observations: List[Union[Observation, MultiAgentObservation]],
+        working_directory_path: Path,
+        learning_algorithm: LearningAlgorithmType,
+    ):
+        super().__init__(model_domain, model_domain_path, observations, working_directory_path, learning_algorithm)
+
+    def calculate_preconditions_semantic_performance(
+        self, learned_domain: Domain, learned_domain_path: Path
+    ) -> Tuple[Dict[str, float], Dict[str, float]]:
+        """Calculates the precision recall values of the learned preconditions.
+
+        :param learned_domain: the action model that was learned using the action model learning algorithm
+        :param learned_domain_path: the path to the learned domain.
+        :return: the precision and recall dictionaries.
+        """
+        num_true_positives = defaultdict(int)
+        num_false_negatives = defaultdict(int)
+        num_false_positives = defaultdict(int)
+        self.logger.debug("Starting to calculate the semantic preconditions performance")
+        for observation in self.dataset_observations:
+            possible_ground_actions = self._create_grounded_action_vocabulary(self.model_domain, observation.grounded_objects)
+            observation_objects = observation.grounded_objects
+            for component in observation.components:
+                possible_ground_actions.extend(component.grounded_joint_action.actions)
+                self.logger.info(
+                    f"Calculating the preconditions' ma semantic performance for the action the state - {component.previous_state.serialize()}"
```

Comment:
no need for "ma semantic" say "Calculating.... semantic performance for the agents' actions"